### ISSUE TITLE
Gtk theme is not automatically installed when using XFCE from Debian Testing

### ISSUE THREAD

**Issue by ahayzen** [flatpak/flatpak#3647]
## Linux distribution and version

Debian Testing using XFCE

## Flatpak version

1.6.3

## Description of the problem

Gtk themes are not automatically installed when using XFCE from Debian Testing.

## Steps to reproduce

1) Load a Debian testing VM using XFCE
2) Install a theme set (such as arc-theme) and set to Arc-Darker
3) Install flatpak and configure flathub
4) Install an app such as gedit
5) Notice that org.gtk.Gtk3theme.Arc-Darker is not installed automatically

## Notes

  * gtk-query-settings gtk-theme-name  matches the app id of the theme
  * Using Debian Stable GNOME (with flatpak 1.2.5) and Ubuntu 20.04 GNOME (with flatpak 1.6.3) both work correctly, so far the only difference i've spotted is it fails with XFCE ?

**Comment by TingPing** [comment]
The automatic feature depends on the value of `gsettings get org.gnome.desktop.interface gtk-theme`

**Comment by ahayzen** [comment]
Hmm interesting, on XFCE `gsettings get org.gnome.desktop.interface gtk-theme` returns `Adwaita` whereas `gtk-query-settings gtk-theme-name` returns `Arc-Darker`.  Wonder if this is why `neofetch` says that "Theme: Arc-Darker [gtk2], Adwaita [gtk3]`.

However when I run gtk3 apps natively I do get Arc-Darker (or whichever theme i pick) even though the gsettings isn't set, so is gtk reading the theme from a different place to flatpak ? Seems weird...

**Comment by TingPing** [comment]
flatpaks cannot read `~/.config/gtk-3.0` so it isn't a valuable file to get settings from. Ideally `flatpak` could read from `XSettings` but that adds quite a bit of complication and may not even help in your case.

I'll just close this as its working as expected for now at least.

**Comment by ahayzen** [comment]
Ugh, so you are saying that Flatpak isn't supported on XFCE ?

Not sure why flatpak itself cannot read from `~/.config/gtk-3.0` as this is needed when you are installing or updating a flatpak not running a flatpak.  (note if one manually installs the theme then it works correctly with flatpak apps, it is just the auto install part that fails, which is an issue in flatpak on the host side detecting the current theme).

And closing an issue because it is "working as expected" when it is clearly broken and should be a feature request instead is not useful.

**Comment by TingPing** [comment]
> Not sure why flatpak itself cannot read from ~/.config/gtk-3.0 as this is needed when you are installing or updating a flatpak not running a flatpak. 

Flatpak'd *applications* cannot read from there, thus the value in that file means nothing.

> (note if one manually installs the theme then it works correctly with flatpak apps)

Ok then that means you have an XSettings service running. This is tricky for `flatpak` to read because then it needs to depend on X11 plus XSettings is a complex thing for what we need.

> Ugh, so you are saying that Flatpak isn't supported on XFCE ?

Don't be hyperbolic, flatpak functions entirely fine. One approach is XFCE could write the GSettings key mentioned above, KDE does this now so its not unreasonable for other desktops to do so as well.

**Comment by ahayzen** [comment]
> > Not sure why flatpak itself cannot read from ~/.config/gtk-3.0 as this is needed when you are installing or updating a flatpak not running a flatpak.
> 
> Flatpak'd _applications_ cannot read from there, thus the value in that file means nothing.

Hmm I don't actually seem to have anything in that folder other than bookmarks anyway (wonder where things are being read).  But my point was that maybe flatpak could read the ini file and leave apps to go via XSettings - but seems this folder is not being used anyway, so that's a dead-end.

> 
> > (note if one manually installs the theme then it works correctly with flatpak apps)
> 
> Ok then that means you have an XSettings service running. This is tricky for `flatpak` to read because then it needs to depend on X11 plus XSettings is a complex thing for what we need.
> 

OK, then maybe the current theme should be queried via portals ? As they have integration into the shell / DE and would be able to tell what theme is running.  Eg xdg-desktop-portal-gtk depends on libgtk-3-0 and then would be able to query what is the current theme in the same way as `gtk-query-settings` ? (and the same for xdg-desktop-portal-kde for the Qt world)

> > Ugh, so you are saying that Flatpak isn't supported on XFCE ?
> 
> Don't be hyperbolic, flatpak functions entirely fine. One approach is XFCE could write the GSettings key mentioned above, KDE does this now so its not unreasonable for other desktops to do so as well.


OK sorry, it mostly works, but just seems like no one has done any polish (looks like preferred applications need work contributed to XFCE as well).

Maybe as XFCE is transitioning from gtk2 -> 3 the gsettings key will be something they can set as well. I'll have a look to see if there are any discussions upstream.

Is there any documentation pointing to `gsettings get org.gnome.desktop.interface gtk-theme` being the "correct" way to set themes for gtk apps ? (as the org.gnome.desktop id seems odd for that :-) ).

**Comment by TingPing** [comment]
> OK, then maybe the current theme should be queried via portals ? 

`xdg-desktop-portal-gtk` gets the theme name from the same GSetting I just mentioned. :/

> Is there any documentation pointing to gsettings get org.gnome.desktop.interface gtk-theme being the "correct" way to set themes for gtk apps ? (as the org.gnome.desktop id seems odd for that :-) ).

I very much agree its a bad id. There is no documentation except for the code: https://github.com/GNOME/gtk/blob/4c9357aaef06e30e4804e86286013110938e5ca7/gdk/wayland/gdkdisplay-wayland.c#L1620

**Comment by ahayzen** [comment]
> > OK, then maybe the current theme should be queried via portals ?
> 
> `xdg-desktop-portal-gtk` gets the theme name from the same GSetting I just mentioned. :/
> 

:-)   Maybe as it appears all DEs aren't yet using the gettings key that xdg-desktop-portal-gtk should check what `gtk-query-settings` gives as well (and could even be done on a whitelist for certain DEs).

Then flatpak on the host could ask xdg-desktop-portal what is the current theme when it is trying to determine if any packages need to be auto installed.

But i'll also report this upstream if I can't find any existing work to see if it can be resolved in the next release.

> > Is there any documentation pointing to gsettings get org.gnome.desktop.interface gtk-theme being the "correct" way to set themes for gtk apps ? (as the org.gnome.desktop id seems odd for that :-) ).
> 
> I very much agree its a bad id. There is no documentation except for the code: https://github.com/GNOME/gtk/blob/4c9357aaef06e30e4804e86286013110938e5ca7/gdk/wayland/gdkdisplay-wayland.c#L1620

**Comment by ahayzen** [comment]
@TingPing XFCE have responded in a bug I opened requesting to set gsettings ( https://gitlab.xfce.org/xfce/xfce4-settings/-/issues/179#note_10473 ) they said that flatpak's own code suggests that it should be reading from XSettings ( https://github.com/flatpak/flatpak/blob/master/common/flatpak-utils.c#L596 ).

It appears that you wrote this code in https://github.com/flatpak/flatpak/pull/1181 :-) You also stated in that merge that "an XSettings method shouldn't be terribly hard to add (probably dynamically load xlib to avoid the dep)." - so why can't this be done now ?

Alex also responded saying that "I think having specific toolkit integration at this level is a bit hacky," - could flatpak asking portals for the current system theme be a less "hacky" solution ?

Therefore could you do one of the following:
  * Comment on the XFCE bug with more reasoning as to why they should be setting the value in gsettings (there may be other purposes I'm not aware of)
  * Re-open this bug as a feature request to add reading the XSettings in flatpak-utils
  * Re-open this bug as a feature request for flatpak to ask via xdg-desktop-portal what is the current theme (and in xdg-desktop-portal-gtk implement both gsettings and XSettings lookups)

**Comment by TingPing** [comment]
> It appears that you wrote this code in #1181 :-) You also stated in that merge that "an XSettings method shouldn't be terribly hard to add (probably dynamically load xlib to avoid the dep)." - so why can't this be done now ?

Well I looked into it and XSettings was far less simple than I expected. It could still be done but I don't plan on doing it.

> Alex also responded saying that "I think having specific toolkit integration at this level is a bit hacky," - could flatpak asking portals for the current system theme be a less "hacky" solution ?

Its just moving the problem around; I don't really have a strong opinion but it would turn a single patch to a single project into something for multiple moving pieces to deal with. As-is I don't think `flatpak` ever talks to the portal so thats a new concept.

**Comment by ahayzen** [comment]
For anyone following this issue, a change to make XFCE set gsettings keys has been merged into master https://gitlab.xfce.org/xfce/xfce4-settings/-/issues/179  - hopefully meaning it is in XFCE 4.16
